File: leetcode/94. Binary Tree Inorder Traversal.cc
// Copyright 2017 Qi Wang
// Date: 2017-02-05
// Morris traversal
class Solution {
 public:
  vector<int> inorderTraversal(TreeNode* root) {
    vector<int> result;
    TreeNode* cur = root;
    while (cur != nullptr) {
      if (cur->left == nullptr) {
        result.push_back(cur->val);
        cur = cur->right;
      } else {
        auto prev = cur->left;
        while (prev->right != nullptr && prev->right != cur)
          prev = prev->right;
        if (prev->right == nullptr) {
          prev->right = cur;
          cur = cur->left;
        // prev->right == cur
        } else {
          result.push_back(cur->val);
          prev->right = nullptr;
          cur = cur->right;
        }
      }
    }
    return result;
  }
};

// Date: ...
class Solution 
{
public:
    vector<int> inorderTraversal(TreeNode *root) 
    {
      vector<int> ret;
      gao(root, ret);
      return ret;
    }
private:
  void gao(TreeNode* node, vector<int>& ret)
  {
    if (node != NULL)
    {
      gao(node->left, ret);
      ret.push_back(node->val);
      gao(node->right, ret);
    }
  }
};

File: leetcode/94. Binary Tree Inorder Traversal.js
// Copyright 2017 Qi Wang
// Date: 2017-05-04
const inorderTraversal = root => {
  if (root === null) return [];
  const result = inorderTraversal(root.left);
  // ... : Destructuring assignment.
  //   - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment
  // Assume result = [1, 2], root.val = 3, inorderTraversal(root.right) =
  // [4, 5, 6]. Then we will get result = [1, 2, 3, 4, 5, 6] after the push
  // operation.
  result.push(root.val, ...inorderTraversal(root.right));
  return result;
};
